pep-0727.rst

```diff
@@ -0,0 +1,516 @@
+
+The current standard method of documenting code APIs in Python is using docstrings.
+But there's no standard way to document parameters in docstrings.
+
+There are several pseudo-standards for the format in these docstrings, and new
+pseudo-standards can appear easily: numpy, Google, Keras, reST, etc.
+
+All these formats are some specific syntax inside a string. Because of this, when
+editing those docstrings, editors can't easily provide support for autocompletion,
+inline errors for broken syntax, etc.
+
+Editors don't have a way to support all the possible micro languages in the docstrings
+and show nice user interfaces when developers use libraries with those different
+formats. They could even add support for some of the syntaxes, but probably not all,
+or not completely.
+
+Because the docstring is in a different place in the code than the actual parameters
+and it requires duplication of information (the parameter name) the information about
+a parameter is easily in a place in the code quite far away from the declaration of
+the actual parameter. This means it's easy to refactor a function, remove a parameter,
+and forget to remove its docs. The same happens when adding a new parameter: it's easy
+to forget to add the docstring for it.
+
+Editors can't check or ensure the consistency of the parameters and their docs.
+
+It's not possible to programatically test these docstrings, for example to ensure
+documentation consistency for the parameters across several similar functions, at
+least not without implementing a custom syntax parser.
+
+Some of these previous formats tried to account for the lack of type annotations
+in older Python versions by including typing information in the docstrings,
+but now that information doesn't need to be in docstrings as there is now an official
+syntax for type annotations.
+
+
+Rationale
+=========
+
+This proposal intends to address these shortcomings by extending and complementing the
+information in docstrings, keeping backwards compatibility with existing docstrings,
+and doing it in a way that leverages the Python language and structure, via type
+annotations with ``Annotated``, and a new function in ``typing``.
+
+The reason why this would belong in the standard Python library instead of an
+external package is because although the implementation would be quite trivial,
+the actual power and benefit from it would come from being a standard, so that
+editors and other tools could implement support for it.
+
+This doesn't deprecate current usage of docstrings, it's transparent to common
+developers (library users), and it's only opt-in for library authors that would
+like to adopt it.
+
+
+Specification
+=============
+
+
+``typing.doc``
+--------------
+
+The main proposal is to have a new function ``doc()`` in the ``typing`` module.
+Even though this is not strictly related to the type annotations, it's expected
+to go in ``Annotated`` type annotations, and to interact with type annotations.
```

Comment:
This is unclear. If I write ``a: doc(...)``, is this valid? What if I want to put a ``doc(...)`` call and something else in the second argument of ``Annotated``?